File: Practice/AllDomains/CoreCS/DataStructures/Stacks/SimpleTextEditor.cs
/*	author:@shivkrthakur  */
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
class Solution {
    static void Main(String[] args) {
        /* Enter your code here. Read input from STDIN. Print output to STDOUT. Your class should be named Solution */
        var q = Convert.ToInt32(Console.ReadLine());
        var opStack = new Stack<Tuple<int,string>>();
        var sb = new StringBuilder();
        for(int i = 0; i < q; i++)
        {
            var input = Console.ReadLine().Split();
            var inp = Convert.ToInt32(input[0]);
            switch(inp){
                case 1:
                    var val = input[1];
                    opStack.Push(new Tuple<int,string>(1,val));
                    sb.Append(val);
                    break;
                case 2:
                    int c = Convert.ToInt32(input[1]);
                    int ind = sb.Length - c;
                    var sub = sb.ToString().Substring(ind, c);
                    sb.Remove(ind, c);
                    opStack.Push(new Tuple<int,string>(2,sub));
                    break;
                case 3:
                    int index = Convert.ToInt32(input[1]);
                    Console.WriteLine(sb[index-1]);
                    break;
                case 4:
                    var tup = opStack.Pop();
                    if(tup.Item1 == 1){
                        var len = tup.Item2.Length;
                        var ind4 = sb.Length - len;
                        sb.Remove(ind4, len);
                    }
                    if(tup.Item1 == 2){
                        sb.Append(tup.Item2);
                    }
                    break;
            }
        }
        
    }
}

File: Practice/AllDomains/CoreCS/DataStructures/Stacks/SimpleTextEditor.js
/*	author:@shivkrthakur  */
var index = 0;

function readLine(input){
    return input[index++];
}

function processData(input) {
    //Enter your code here
    var q = parseInt(readLine(input));
    var sb = '';
    var stack = [];
    for(var i = 0; i < q; i++){
        var val = readLine(input).split(' ');
        var op = parseInt(val[0]);
        switch(op){
          case 1:
            var str = val[1];
            sb += str;
            stack.push(val);
            break;
          case 2:
            var c = parseInt(val[1]);
            stack.push([op, sb.substring(sb.length - c, sb.length)]);
            sb = sb.substring(0,sb.length - c);
            break;
          case 3:
            var index = parseInt(val[1]);
            console.log(sb[index - 1]);
            break;
          case 4:
            var p = stack.pop();
            if(p[0] == '1'){
              var s1 = p[1];
              sb = sb.substring(0, sb.length - s1.length);
            }
            if(p[0] == '2'){
              sb += p[1];
            }
            break;
        }
    }
} 

process.stdin.resume();
process.stdin.setEncoding("ascii");
_input = "";
process.stdin.on("data", function (input) {
    _input += input;
});

process.stdin.on("end", function () {
   processData(_input.split('\n'));
});
